Combat.

BautistaDalmasso opened the pull request:
My attempt at helping Fyx with finishing combat.

BautistaDalmasso committed:
Big reformat to actions.

The `Action.action` function now uses `_action_with_target`, `_perform_action`
and `_no_target_action` to separate the logic of performing actions.
`action` func is now able to handle an action with a single target.

AbooMinister25 commented on server/game_components/game_objects.py:
```diff
@@ -240,87 +240,68 @@ def action(self, _caster: Entity, _target: Entity | None) -> list[ActionDict]:
         :return: its a list of what happened. Single target attacks still return a list of dicts
         """
-        if _target is None:
-            assert self.requires_target is False
```
you could shorten this to `assert not self.requires_target`. Also, consider not using `assert` here, and raise an exception instead (it can be a custom one).
```py
if self.requres_target:
    raise SomeError("blah blah")
``` 
`assert` is primarily a tool meant for use in tests, and in cases where some condition should never happen. Python will ignore `assert`s and won't compile them when you run the program with optimizations on and/or when `__debug__` is set to `False`.

AbooMinister25 commented on server/game_components/game_objects.py:
```diff
@@ -240,87 +240,68 @@ def action(self, _caster: Entity, _target: Entity | None) -> list[ActionDict]:
-                        "cast": cast,
-                    }
-                ]
+            assert self.requires_target is False
```
Same thing as above

AbooMinister25 commented on server/game_components/game_objects.py:
```diff
@@ -240,87 +240,68 @@ def action(self, _caster: Entity, _target: Entity | None) -> list[ActionDict]:
+            assert self.requires_target is False
+            action_list = self._no_target_action(cast, _caster)
+        else:
+            assert self.requires_target is True
```
this could be shortened to `assert self.requires_target`, on top of my other comments above.

BautistaDalmasso committed:
Reformat to player updating and command queueing.

Actions can now be added to the player's queue if they are valid.
When player.update is used the player now executes it's next action in it's
command queue.

BautistaDalmasso committed:
Reformat `Game.update`.

BautistaDalmasso committed:
Allows a player's queued movement to be executed.

BautistaDalmasso committed:
Action requests are now added to the player's action queue.

BautistaDalmasso committed:
Allows the player to send movement requests.

The client can now do `/move [direction]`.
The server will receive the movement request, scramble it and add the
scrambled action to the players queue!

BautistaDalmasso committed:
Server now properly responds to bad commands.

If the player tries to make a command that isn't available the server
will send a proper response.
Removes the `assert` that threw an error when the server received a bad
command.

BautistaDalmasso committed:
Merge branch 'main' into combat-bauti

AbooMinister25 reviewed (approved):
LGTM
